packages/jupyter-ai/jupyter_ai/config_manager.py

```diff
@@ -12,120 +16,317 @@
+        modified after this time, then we can return the cached
+        `self._config`."""
+        self._last_read: Optional[int] = None
 
-    def get_lm_provider(self):
-        return self.lm_provider
+        """In-memory cache of the `GlobalConfig` object parsed from the config
+        file."""
+        self._config: Optional[GlobalConfig] = None
 
-    def get_lm_provider_params(self):
-        return self.lm_provider_params
+        self._init_config_schema()
+        self._init_validator()
+        self._init_config()
 
-    def get_em_provider(self):
-        return self.em_provider
+    def _init_config_schema(self):
+        if not os.path.exists(self.schema_path):
+            os.makedirs(os.path.dirname(self.schema_path), exist_ok=True)
+            shutil.copy(OUR_SCHEMA_PATH, self.schema_path)
 
-    def get_em_provider_params(self):
-        return self.em_provider_params
+    def _init_validator(self) -> Validator:
+        with open(OUR_SCHEMA_PATH, encoding="utf-8") as f:
+            schema = json.loads(f.read())
+            Validator.check_schema(schema)
+            self.validator = Validator(schema)
 
-    def _authenticate_provider(
-        self,
-        provider: AnyProvider,
-        provider_params: Dict[str, Any],
-        config: GlobalConfig,
-    ):
-        auth_strategy = provider.auth_strategy
-        if auth_strategy and auth_strategy.type == "env":
-            api_keys = config.api_keys
-            name = auth_strategy.name
-            if name not in api_keys:
+    def _init_config(self):
+        if os.path.exists(self.config_path):
+            with open(self.config_path, encoding="utf-8") as f:
+                config = GlobalConfig(**json.loads(f.read()))
+                # re-write to the file to validate the config and apply any
+                # updates to the config file immediately
+                self._write_config(config)
+            return
+
+        properties = self.validator.schema.get("properties", {})
+        field_list = GlobalConfig.__fields__.keys()
+        field_dict = {
+            field: properties.get(field).get("default") for field in field_list
+        }
+        default_config = GlobalConfig(**field_dict)
+        self._write_config(default_config)
+
+    def _read_config(self) -> GlobalConfig:
+        """Returns the user's current configuration as a GlobalConfig object.
+        This should never be sent to the client as it includes API keys. Prefer
+        self.get_config() for sending the config to the client."""
+        if self._config and self._last_read:
+            last_write = os.stat(self.config_path).st_mtime_ns
+            if last_write <= self._last_read:
+                return self._config
+
+        with open(self.config_path, encoding="utf-8") as f:
+            self._last_read = time.time_ns()
+            raw_config = json.loads(f.read())
+            config = GlobalConfig(**raw_config)
+            self._validate_config(config)
```

Comment:
Can we catch the `ValueError` here, log something to the console, and pick a default language or embedding model, if the selected model is not available? This would mitigate #346, where the chat UI dies entirely if the user had previously picked a model that is no longer offered.

Reply:
Yes, #346 is a pretty serious bug. Let me see if I can fix it here as well.

Reply:
Upon further investigation, it seems like #346 isn't actually caused by an invalid model being specified in the config, but due to some other reason. Given that no users have encountered this edge case (as it only happens when you manually alter the config.json file), I think it's safe to ignore this for now.